core/src/sumeragi/main_loop.rs

```diff
@@ -517,90 +505,82 @@ fn handle_message(
         (Message::ViewChangeSuggested, _) => {
             trace!("Received view change suggestion.");
         }
-        (Message::BlockSyncUpdate(block_sync_update), _) => {
-            let block_hash = block_sync_update.hash();
-            info!(%addr, %role, hash=%block_hash, "Block sync update received");
-
-            let block = match block_sync_update
-                .clone()
-                .validate_and_extract_block::<false>(
-                    &sumeragi.transaction_validator,
-                    state.wsv.clone(),
-                    state.latest_block_hash,
-                    state.latest_block_height,
-                )
-                .or_else(|_|
-                /* If the block fails validation we must check again using the finaziled wsv.
-                When a soft-fork occurs the consensus-block may be valid on the previous
-                wsv but not the current one. */
-                block_sync_update.validate_and_extract_block::<false>(
-                    &sumeragi.transaction_validator,
-                    state.finalized_wsv.clone(),
-                    state.previous_block_hash,
-                    state.latest_block_height.saturating_sub(1),
-                )) {
+        (Message::BlockSyncUpdate(BlockSyncUpdate { block }), _) => {
+            let block_hash = block.hash();
+            debug!(%addr, %role, %block_hash, "Block sync update received");
+
+            let block = match block.validate_without_validating_topology(
+                state.latest_block_height,
+                state.latest_block_hash,
+                &sumeragi.transaction_validator,
+                state.wsv.clone(),
+            ) {
                 Ok(block) => block,
-                Err(error) => {
-                    error!(%addr, %role, %block_hash, ?error, "Block not valid.");
-                    return;
+                Err((block, error)) => {
+                    let header = block.header();
+
+                    // TODO: What is the correct condition for soft-fork?
```

Comment:
I think correct condition is when:
1) block pass validation against `finalized_wsv` (this will include height check, hash check)
2) block `view_change_index` is greater then currently committed one.

Reply:
I can confirm this. When two blocks fit the same slot in the chain, that is a soft-fork and we revert and use the one with the highest view change index.